mm2src/coins/utxo/rpc_clients/connection_manager_multiple.rs

```diff
@@ -0,0 +1,402 @@
+use async_trait::async_trait;
+use core::time::Duration;
+use futures::lock::{Mutex as AsyncMutex, MutexGuard};
+use futures::{select, FutureExt};
+use std::collections::HashMap;
+use std::ops::Deref;
+use std::sync::Arc;
+
+use crate::utxo::ScripthashNotificationSender;
+use common::executor::abortable_queue::{AbortableQueue, WeakSpawner};
+use common::executor::{AbortableSystem, SpawnFuture, Timer};
+use common::log::{debug, error, info, warn};
+
+use super::connection_manager_common::{ConnectionManagerErr, ConnectionManagerTrait, DEFAULT_CONN_TIMEOUT_SEC,
+                                       SUSPEND_TIMEOUT_INIT_SEC};
+use super::{spawn_electrum, ElectrumClientEvent};
+use super::{ElectrumConnSettings, ElectrumConnection};
+
+#[derive(Clone, Debug)]
+pub struct ConnectionManagerMultiple(pub Arc<ConnectionManagerMultipleImpl>);
+
+#[derive(Debug)]
+pub struct ConnectionManagerMultipleImpl {
+    guarded: AsyncMutex<ConnectionManagerMultipleState>,
+    abortable_system: AbortableQueue,
+    event_sender: futures::channel::mpsc::UnboundedSender<ElectrumClientEvent>,
+    scripthash_notification_sender: ScripthashNotificationSender,
+}
+
+#[derive(Debug)]
+struct ConnectionManagerMultipleState {
+    connection_contexts: Vec<ElectrumConnCtx>,
+    scripthash_subs: HashMap<String, Arc<AsyncMutex<ElectrumConnection>>>,
+}
+
+#[derive(Debug)]
+struct ElectrumConnCtx {
+    conn_settings: ElectrumConnSettings,
+    abortable_system: AbortableQueue,
+    suspend_timeout_sec: u64,
+    connection: Option<Arc<AsyncMutex<ElectrumConnection>>>,
+}
+
+#[async_trait]
+impl ConnectionManagerTrait for ConnectionManagerMultiple {
+    async fn get_connection(&self) -> Vec<Arc<AsyncMutex<ElectrumConnection>>> { self.0.get_connection().await }
+
+    async fn get_connection_by_address(
+        &self,
+        address: &str,
+    ) -> Result<Arc<AsyncMutex<ElectrumConnection>>, ConnectionManagerErr> {
+        self.0.get_connection_by_address(address).await
+    }
+
+    async fn connect(&self) -> Result<(), ConnectionManagerErr> { self.deref().connect().await }
```

Comment:
This should be an infinite recursive call.

Reply:
moved impl login to connect in `impl ConnectionManagerTrait for ConnectionManagerMultiple`

Reply:
Cool, should work. I am wondering this might be why some tests are failing.